File: Backend.NET/Controllers/BookingController.cs
using Microsoft.AspNetCore.Mvc;
using System.Collections.Generic;
using FleetManagementSystem.Api.DTOs;
using FleetManagementSystem.Api.Services;
using Microsoft.Extensions.Logging;

namespace FleetManagementSystem.Api.Controllers;

[ApiController]
[Route("booking")]
public class BookingController : ControllerBase
{
    private readonly IBookingService _bookingService;
    private readonly ILogger<BookingController> _logger;

    public BookingController(IBookingService bookingService, ILogger<BookingController> logger)
    {
        _bookingService = bookingService;
        _logger = logger;
    }

    [HttpGet("heartbeat")]
    public IActionResult Heartbeat() => Ok("Booking Controller is Running - Version 3 (Get Fix)");

    [HttpGet("user/{email}")]
    public async Task<ActionResult<List<BookingResponse>>> GetBookingsByUser(string email)
    {
        return Ok(await _bookingService.GetBookingsByEmail(email));
    }

    [HttpGet("all")]
    public async Task<ActionResult<List<BookingResponse>>> GetAllBookings()
    {
        return Ok(await _bookingService.GetAllBookings());
    }

    [HttpGet("hub/{hubId}")]
    public async Task<ActionResult<List<BookingResponse>>> GetBookingsByHub(int hubId)
    {
        return Ok(await _bookingService.GetBookingsByHub(hubId));
    }

    [HttpPost("create")]
    public async Task<ActionResult<BookingResponse>> CreateBooking([FromBody] BookingRequest request)
    {
        _logger.LogInformation("Creating booking for customer: {Email}, CarId: {CarId}", request.Email, request.CarId);
        try
        {
            return Ok(await _bookingService.CreateBooking(request));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating booking for customer: {Email}", request.Email);
            return BadRequest(ex.Message);
        }
    }

    [HttpPost("handover/{id}")]
    public async Task<ActionResult<BookingResponse>> HandoverCar([FromRoute] long id)
    {
        var req = new HandoverRequest { BookingId = id };
        try
        {
            return Ok(await _bookingService.ProcessHandover(req));
        }
        catch(Exception ex)
        {
            _logger.LogError(ex, "Error processing handover for booking ID: {Id}", id);
            return BadRequest(ex.Message);
        }
    }

    [HttpPost("process-handover")]
    public async Task<ActionResult<BookingResponse>> ProcessHandover([FromBody] HandoverRequest request)
    {
        try
        {
            return Ok(await _bookingService.ProcessHandover(request));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing handover for booking ID: {Id}", request.BookingId);
            return BadRequest(ex.Message);
        }
    }

    [HttpPost("return")]
    public async Task<ActionResult<BookingResponse>> ReturnCar([FromBody] ReturnRequest request)
    {
        try
        {
            return Ok(await _bookingService.ReturnCar(request));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing return for booking ID: {Id}", request.BookingId);
            return BadRequest(ex.Message);
        }
    }

    [HttpPost("cancel/{id}")]
    public async Task<ActionResult<BookingResponse>> CancelBooking([FromRoute] long id)
    {
        try
        {
            return Ok(await _bookingService.CancelBooking(id));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error cancelling booking ID: {Id}", id);
            return BadRequest(ex.Message);
        }
    }

    [HttpPut("modify/{id}")]
    public async Task<ActionResult<BookingResponse>> ModifyBooking([FromRoute] long id, [FromBody] BookingRequest request)
    {
        _logger.LogInformation("Modifying booking ID: {Id}", id);
        try
        {
             return Ok(await _bookingService.ModifyBooking(id, request));
        }
        catch (Exception ex)
        {
             _logger.LogError(ex, "Error modifying booking ID: {Id}", id);
             return BadRequest(ex.Message);
        }
    }

    [HttpPost("storeDates")]
    public ActionResult<string> StoreBookingDates([FromQuery] string start_date, [FromQuery] string end_date, [FromQuery] int cust_id)
    {
        return Ok("success");
    }

    [HttpGet("get/{id}")]
    public async Task<ActionResult<BookingResponse>> GetBooking([FromRoute] string id)
    {
        try
        {
            var response = await _bookingService.GetBooking(id);
            return Ok(response);
        }
        catch (Exception ex)
        {
            return NotFound(ex.Message);
        }
    }
}

File: Backend.JAVA/src/main/java/com/example/demo/controller/BookingController.java
package com.example.demo.controller;

import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import com.example.demo.Service.BookingService;
import com.example.demo.dto.BookingRequest;
import com.example.demo.dto.BookingResponse;
import com.example.demo.dto.ReturnRequest;
import com.example.demo.dto.HandoverRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import java.time.format.DateTimeFormatter;

@RestController
@RequestMapping("/booking")
@CrossOrigin(origins = "http://localhost:3000")
public class BookingController {

    @Autowired
    private BookingService bookingService;

    @GetMapping("/get/{bookingId}") // Alias to match frontend
    public ResponseEntity<BookingResponse> getBooking(@PathVariable Long bookingId) {
        BookingResponse booking = bookingService.getBooking(bookingId);
        return ResponseEntity.ok(booking);
    }

    @GetMapping("/{bookingId}") // Keep original for backward compatibility? Or just alias.
    public ResponseEntity<BookingResponse> getBookingById(@PathVariable Long bookingId) {
        return getBooking(bookingId);
    }

    @GetMapping("/hub/{hubId}")
    public ResponseEntity<java.util.List<BookingResponse>> getBookingsByHub(@PathVariable Integer hubId) {
        java.util.List<BookingResponse> bookings = bookingService.getBookingsByHubId(hubId);
        return ResponseEntity.ok(bookings);
    }

    @GetMapping("/user/{email}")
    public ResponseEntity<java.util.List<BookingResponse>> getBookingsByUser(@PathVariable String email) {
        java.util.List<BookingResponse> bookings = bookingService.getBookingsByEmail(email);
        return ResponseEntity.ok(bookings);
    }

    @GetMapping("/all")
    public ResponseEntity<java.util.List<BookingResponse>> getAllBookings() {
        return ResponseEntity.ok(bookingService.getAllBookings());
    }

    @PostMapping("/create")
    public ResponseEntity<BookingResponse> createBooking(@RequestBody BookingRequest request) {
        BookingResponse booking = bookingService.createBooking(request);
        return ResponseEntity.ok(booking);
    }

    @PostMapping("/handover/{bookingId}")
    public ResponseEntity<BookingResponse> handoverCar(@PathVariable Long bookingId) {
        BookingResponse booking = bookingService.handoverCar(bookingId);
        return ResponseEntity.ok(booking);
    }

    @PostMapping("/process-handover")
    public ResponseEntity<BookingResponse> processHandover(@RequestBody HandoverRequest request) {
        BookingResponse booking = bookingService.processHandover(request);
        return ResponseEntity.ok(booking);
    }

    @PostMapping("/return")
    public ResponseEntity<BookingResponse> returnCar(@RequestBody ReturnRequest request) {
        BookingResponse booking = bookingService.returnCar(request);
        return ResponseEntity.ok(booking);
    }

    // Endpoint removed as per single-integer ID refactor
    /*
     * @GetMapping("/search/{confirmationNumber}")
     * public ResponseEntity<BookingResponse> getBookingByConfirmation(@PathVariable
     * String confirmationNumber) {
     * BookingResponse booking =
     * bookingService.getBookingByConfirmation(confirmationNumber);
     * return ResponseEntity.ok(booking);
     * }
     */

    @PostMapping("/cancel/{bookingId}")
    public ResponseEntity<BookingResponse> cancelBooking(@PathVariable Long bookingId) {
        BookingResponse booking = bookingService.cancelBooking(bookingId);
        return ResponseEntity.ok(booking);
    }

    @PutMapping("/modify/{bookingId}")
    public ResponseEntity<BookingResponse> modifyBooking(@PathVariable Long bookingId,
            @RequestBody BookingRequest request) {
        BookingResponse booking = bookingService.modifyBooking(bookingId, request);
        return ResponseEntity.ok(booking);
    }

    @PostMapping("/storeDates")
    public String storeBookingDates(@RequestParam String start_date, @RequestParam String end_date,
            @RequestParam int cust_id) {

        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
        LocalDateTime rentalDate = LocalDateTime.parse(start_date, dateTimeFormatter);
        LocalDateTime returnDate = LocalDateTime.parse(end_date, dateTimeFormatter);
        //
        // return
        // initialDateService.storeTempDateandTime(rentalDate,returnDate,cust_id);
        return "success";
    }

}
